### PR TITLE
Add knowledge distillation example

### PR DESCRIPTION
**Pull request by Kennethborup**

### CODE REVIEW

File: examples/vision/knowledgedistillation.py [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        # Update weights
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        # Update the metrics configured in `compile()`.
+        self.compiled_metrics.update_state(y, s_pred)
+
+        # Return a dict of performance
+        return {
+            **{"s_loss": s_loss, "d_loss": d_loss},
+            **{m.name: m.result() for m in self.metrics},
```
Comment (fchollet) [inline examples/vision/knowledgedistillation.py:99]: More readable:

```
results = {m.name: m.result() for m in self.metrics}
results.update({"s_loss": s_loss, "d_loss": d_loss})
return results
```

File: examples/vision/knowledgedistillation.py [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+"""
+
+
+class Distiller(keras.Model):
+    def __init__(self, student, teacher):
+        super(Distiller, self).__init__()
+        self.teacher = student
+        self.student = teacher
+
+    def compile(self, optimizer, metrics, s_loss_fn, d_loss_fn, alpha, T):
```
Comment (fchollet) [inline examples/vision/knowledgedistillation.py:60]: Please add a docstring with argument descriptions.

Prefer fully-spelled out lower case snake case variable names, e.g. `temperature` and `student_loss`.

`alpha` and `temperature` are arguments that should have a default, e.g. 0.1 and 3.

File: examples/vision/knowledgedistillation.py [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        layers.Flatten(),
+        layers.Dense(10),
+    ],
+    name="student",
+)
+
+"""
+## Prepare dataset
+The dataset used for training the teacher and distilling the teacher is
+[MNIST](https://keras.io/api/datasets/mnist/). Both the student and teacher are trained
```
Comment (fchollet) [inline examples/vision/knowledgedistillation.py:157]: Could we use CIFAR10, and accordingly larger convnets?

File: examples/vision/knowledgedistillation.py [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+
+# Normalize data
+x_train = x_train.astype("float32") / 255.0
+x_train = np.reshape(x_train, (-1, 28, 28, 1))
+
+x_test = x_test.astype("float32") / 255.0
+x_test = np.reshape(x_test, (-1, 28, 28, 1))
+
+# Construct datasets
+train_dataset = tf.data.Dataset.from_tensor_slices((x_train, y_train))
```
Comment (fchollet) [inline examples/vision/knowledgedistillation.py:173]: Since your data is already NumPy arrays, you don't need to create datasets, just pass the arrays directly to `fit`

File: examples/vision/knowledgedistillation.py [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+
+"""
+## Train teacher
+In knowledge distillation we assume that the teacher is trained and fixed. Thus, we start
+by training the teacher model on the training set in the usual way.
+"""
+
+# Train teacher as usual
+teacher.compile(
+    optimizer="adam",
```
Comment (fchollet) [inline examples/vision/knowledgedistillation.py:187]: Here and in the `compile` call below, since you pass losses as objects, for consistency prefer also passing optimizer and metrics as objects.

File: examples/vision/knowledgedistillation.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+# Distill teacher to student
+distiller.fit(train_dataset.take(10), epochs=1)
+
+# Evaluate student on test dataset
+accuracy = distiller.evaluate(test_dataset.take(10))
+
+"""
+If the teacher is trained for 5 full epochs and the student is distilled on this teacher
+for 5 epochs, you should experience a performance boost compared to training the same
```
Comment (fchollet) [inline examples/vision/knowledgedistillation.py:222]: Is this the case with this example? What are the numbers?

File: examples/vision/knowledgedistillation.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+Knowledge Distillation (or student-teacher training) is a procedure for model
+compression, in which a small (student) model is trained to match a large pre-trained
+(teacher) model. Intrinsic knowledge is transferred from the teacher model to the student
+by minimizing a loss function, aimed at matching softened teacher logits as well as
+ground-truth labels.
+The logits are softened by applying a temperature in the softmax, effectively smoothing
+out the probability distribution and revealing inter-class relationships learned by the
+teacher.
+
+- [Paper](https://arxiv.org/abs/1503.02531)
```
Comment (fchollet) [inline examples/vision/knowledgedistillation.py:21]:
```
**Reference:**

- [Hinton et al, 2015](...)
```

File: examples/vision/knowledgedistillation.py [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+# Train teacher as usual
+teacher.compile(
+    optimizer="adam",
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    metrics=["accuracy"],
+)
+
+# Train and evaluate on subset of data to reduce computations.
+teacher.fit(train_dataset.take(10), epochs=1)
+loss, accuracy = teacher.evaluate(test_dataset.take(10))
```
Comment (fchollet) [inline examples/vision/knowledgedistillation.py:194]: Print a sentence to display the results

File: examples/vision/knowledgedistillation.py [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    d_loss_fn=keras.losses.KLDivergence(),
+    alpha=0.1,
+    T=3,
+)
+
+# Distill teacher to student
+distiller.fit(train_dataset.take(10), epochs=1)
+
+# Evaluate student on test dataset
+accuracy = distiller.evaluate(test_dataset.take(10))
```
Comment (fchollet) [inline examples/vision/knowledgedistillation.py:218]: Print a sentence to display the results

File: examples/vision/knowledgedistillation.py (hunk 8 above)
Comment (fchollet) [reply examples/vision/knowledgedistillation.py:194]: Rather -- just call `model.evaluate()`, it already prints results.

File: examples/vision/knowledgedistillation.py (hunk 9 above)
Comment (fchollet) [reply examples/vision/knowledgedistillation.py:218]: Likewise -- just call `model.evaluate()`, it already prints results.

File: examples/vision/knowledgedistillation.py (hunk 3 above)
Comment (Kennethborup) [reply examples/vision/knowledgedistillation.py:157]: Yes, this example should be compatible with most classical image classification tasks (given the right dataset and models), and minor adjustments could adapt it to a variety of other model types.

I have added a note about this in the example.

File: examples/vision/knowledgedistillation.py (hunk 4 above)
Comment (Kennethborup) [reply examples/vision/knowledgedistillation.py:173]: I use the dataset structure in order to reduce the amount of batches for training in the example in a concise manner. I am unsure of a pretty way to do this if we merely pass in the NumPy arrays.
It would be something like `.fit(x_train[:640], y_train[:640], batch_size=64, epochs=1)` compared to `.fit(train_dataset.take(10), epochs=1)`.

File: examples/vision/knowledgedistillation.py (hunk 6 above)
Comment (Kennethborup) [reply examples/vision/knowledgedistillation.py:222]: I have added an additional section, where the student model is trained from scratch without knowledge distillation, along with performance metrics when training on all data. With a fixed seed, it amounts to approx.:

- Teacher: 97.6%
- Distillation student: 98.1%
- Classical student: 97.6%

**Comment by Kennethborup** [comment]
Thanks for the code review! I have made the required changes, and added a short section in the end for comparing the performance of the student, when trained with and without knowledge distillation.

File: examples/vision/knowledge_distillation.py [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+
+# Train teacher as usual
+teacher.compile(
+    optimizer=keras.optimizers.Adam(),
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    metrics=[keras.metrics.SparseCategoricalAccuracy()],
+)
+
+# Train and evaluate on subset of data to reduce computations.
+teacher.fit(train_dataset.take(10), epochs=5)
```
Comment (fchollet) [inline examples/vision/knowledge_distillation.py:225]: Please train on the full dataset. It is a small model and a small dataset. It's only worth using a subset of the data if it takes 10+ min on GPU, which is not the case here.

That way you can also stick to using NumPy arrays for training.

File: examples/vision/knowledge_distillation.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+import numpy as np
+
+tf.random.set_seed(0)
+np.random.seed(0)
+
+"""
+## Construct `Distiller()` class
```
Comment (fchollet) [inline examples/vision/knowledge_distillation.py:39]: Please aerate the text a bit: line break after titles, line break before bullet points list (important to ensure proper markdown rendering)

**Review by fchollet — COMMENTED** [review]
Looks good, thank you for the update!

**Comment by Kennethborup** [comment]
I have added the updates - thank you for the reviews!

**Review by fchollet — APPROVED** [review]
LGTM -- Thank you for the great contribution!
